l1-contracts/contracts/bridge/L1SharedBridge.sol

```diff
@@ -0,0 +1,650 @@
+        }
+
+        // Withdraw funds
+        if (_l1Token == ETH_TOKEN_ADDRESS) {
+            bool callSuccess;
+            // Low-level assembly call, to avoid any memory copying (save gas)
+            assembly {
+                callSuccess := call(gas(), _depositSender, _amount, 0, 0, 0, 0)
+            }
+            require(callSuccess, "ShB: claimFailedDeposit failed");
+        } else {
+            IERC20(_l1Token).safeTransfer(_depositSender, _amount);
+            // Note we don't allow weth deposits anymore, but there might be legacy weth deposits.
+            // until we add Weth bridging capabilities, we don't wrap/unwrap weth to ether.
+        }
+
+        emit ClaimedFailedDepositSharedBridge(_chainId, _depositSender, _l1Token, _amount);
+    }
+
+    function _isLegacyWithdrawal(uint256 _chainId, uint256 _l2BatchNumber) internal view returns (bool) {
+        return (_chainId == ERA_CHAIN_ID) && (_l2BatchNumber < eraIsEthWithdrawalFinalizedStorageSwitchBatchNumber);
+    }
+
+    /// @notice Finalize the withdrawal and release funds
+    /// @param _l2BatchNumber The L2 batch number where the withdrawal was processed
+    /// @param _l2MessageIndex The position in the L2 logs Merkle tree of the l2Log that was sent with the message
+    /// @param _l2TxNumberInBatch The L2 transaction number in the batch, in which the log was sent
+    /// @param _message The L2 withdraw data, stored in an L2 -> L1 message
+    /// @param _merkleProof The Merkle proof of the inclusion L2 -> L1 message about withdrawal initialization
+    function finalizeWithdrawal(
+        uint256 _chainId,
+        uint256 _l2BatchNumber,
+        uint256 _l2MessageIndex,
+        uint16 _l2TxNumberInBatch,
+        bytes calldata _message,
+        bytes32[] calldata _merkleProof
+    ) external override {
+        // To avoid rewithdrawing txs that have already happened on the legacy bridge
+        // note: new withdraws are all recorded here, so double withdrawing them is not possible
+        bool legacyWithdrawal = _isLegacyWithdrawal(_chainId, _l2BatchNumber);
+        if (legacyWithdrawal) {
+            require(!legacyBridge.isWithdrawalFinalized(_l2BatchNumber, _l2MessageIndex), "ShB: legacy withdrawal");
+        }
+        _finalizeWithdrawal(_chainId, _l2BatchNumber, _l2MessageIndex, _l2TxNumberInBatch, _message, _merkleProof);
+    }
+
+    function finalizeWithdrawalLegacyErc20Bridge(
+        uint256 _l2BatchNumber,
+        uint256 _l2MessageIndex,
+        uint16 _l2TxNumberInBatch,
+        bytes calldata _message,
+        bytes32[] calldata _merkleProof
+    ) external override onlyLegacyBridge returns (address l1Receiver, address l1Token, uint256 amount) {
+        return
+            _finalizeWithdrawal(
+                ERA_CHAIN_ID,
+                _l2BatchNumber,
+                _l2MessageIndex,
+                _l2TxNumberInBatch,
+                _message,
+                _merkleProof
+            );
```

Comment:
NIT: mix of usual and names return

Reply:
Yes but here I think it makes sense to know what the values are that are returned, but I can add comments instead